### PR TITLE
Add support for the OAUTHBEARER mechnism

### PR DESCRIPTION
**Pull request by frodeaa**
fixes: https://github.com/emersion/go-sasl/issues/6

### CODE REVIEW

File: oauthbearer.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	Schemes string `json:"schemes"`
+	Scope   string `json:"scope"`
+}
+
+// Implements error
+func (err *OAuthBearerError) Error() string {
+	return fmt.Sprintf("OAUTHBEARER authentication error (%v)", err.Status)
+}
+
+type oauthbearerClient struct {
```
Comment (emersion) [inline oauthbearer.go]: Style: this should be `oauthBearerClient`

File: oauthbearer.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	authBearerErr := &OAuthBearerError{}
+	if err := json.Unmarshal(challenge, authBearerErr); err != nil {
+		return nil, err
+	} else {
+		return nil, authBearerErr
+	}
+}
+
+// An implementation of the OAUTHBEARER authentication mechanism, as
+// described in https://tools.ietf.org/html/rfc7628
```
Comment (emersion) [inline oauthbearer.go]: We can just write "RFC 7628" instead of the link, GoDoc will automatically linkify it.

File: oauthbearer.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+func NewOAuthBearerClient(username, token string) Client {
```
Comment (emersion) [inline oauthbearer.go]: This is missing host and port. Maybe a struct should be used instead of separate parameters?

**Comment by frodeaa** [comment]
> Some tests would be nice to have (could be extracted from https://tools.ietf.org/html/rfc7628#section-4)

ok

File: oauthbearer.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+type oauthBearerClient struct {
+	OAuthBearerOptions
+}
+
+func (a *oauthBearerClient) Start() (mech string, ir []byte, err error) {
+	mech = OAuthBearer
+	ir = []byte("n,a=" + a.Username +
+		",\x01host=" + a.Host +
+		"\x01port=" + strconv.Itoa(a.Port) +
```
Comment (emersion) [inline oauthbearer.go]: Question: are these fields required? Does Gmail just ignore these?

Comment (frodeaa) [reply oauthbearer.go]: Gmail ignores host and port but fails if a={Username} does not match the token

> Unknown key/value pairs MUST be ignored by the
   server.
https://tools.ietf.org/html/rfc7628#section-3.1

so another question is if there is a case for supporting "any" fields?

File: oauthbearer.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+type oauthBearerClient struct {
+	OAuthBearerOptions
+}
+
+func (a *oauthBearerClient) Start() (mech string, ir []byte, err error) {
+	mech = OAuthBearer
+	ir = []byte("n,a=" + a.Username +
+		",\x01host=" + a.Host +
```
Comment (emersion) [inline oauthbearer.go]: We should probably only append host and port if they aren't zero values

File: oauthbearer.go (hunk 4 above)
Comment (emersion) [reply oauthbearer.go]: Okay.

I think we should just add fields to the struct if an RFC extends OAUTHBEARER.

**Comment by emersion** [comment]
Thanks for adding the tests!

**Pull request merged**